File: vicon/src/cpp_swarm/cpp_swarm/cpp_uav.py
#!/bin/python3

"""
SCRIPT THAT PERFORMS COVERAGE PATH PLANNING OF 1 DRONE
"""

import rclpy
from rclpy.node import Node
from rclpy.clock import Clock
import numpy as np

from px4_msgs.msg import OffboardControlMode
from px4_msgs.msg import TrajectorySetpoint
from px4_msgs.msg import VehicleCommand
from px4_msgs.msg import VehicleStatus

from rclpy.qos import (
    QoSProfile, 
    QoSReliabilityPolicy, 
    QoSHistoryPolicy)

class OffboardControl(Node):

    def __init__(self):
        super().__init__('OffboardControl')
        self.offboard_control_mode_publisher_ = self.create_publisher(OffboardControlMode,"/fmu/in/offboard_control_mode", 10)
        self.trajectory_setpoint_publisher_ = self.create_publisher(TrajectorySetpoint,"/fmu/in/trajectory_setpoint", 10)
        self.vehicle_command_publisher_ = self.create_publisher(VehicleCommand,"/fmu/in/vehicle_command", 10)
        self.vehicle_status_publisher_ = self.create_publisher(VehicleStatus,"/fmu/in/vehicle_status", 10)

        # Stato missione per macchina a stati
        self.mission_state = 0
        # Waypoint corrente pubblicato da timer_offboard_cb()
        self.current_waypoint = [0.0, 0.0, 0.0]

        # Timers
        self.timer_offboard = self.create_timer(0.1, self.timer_offboard_cb)
        self.timer_mission = self.create_timer(10, self.mission)

        qos_profile = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=1
        )

    def mission(self):

        if self.mission_state == 0:
            """ Arming and takeoff to 1st waypoint """
            self.get_logger().info("Mission started")
            self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_DO_SET_MODE, 1., 6.)
            self.mission_state = 1

        elif self.mission_state == 1:    
            self.arm()
            self.get_logger().info("Vehicle armed")
          
            self.get_logger().info("First waypoint")
            self.current_waypoint = np.asfarray([-2.5, -2.5, 2], dtype = np.float32)
            self.ENU2NED_vector_converter()
            print(self.current_waypoint)
            self.mission_state = 2

        elif self.mission_state == 2:
            """Waypoint 2"""
            self.get_logger().info("Second waypoint")
            # Imposta secondo waypoint, 
            self.current_waypoint = np.asfarray([-2.5, 2.5, 2], dtype = np.float32)
            self.ENU2NED_vector_converter()
            print(self.current_waypoint)
            
            self.mission_state = 3
        
        elif self.mission_state == 3:
            """Waypoint 2"""
            self.get_logger().info("Third waypoint")
            # Imposta secondo waypoint
            self.current_waypoint = np.asfarray([2.5, -2.5, 2], dtype = np.float32)
            self.ENU2NED_vector_converter()
            print(self.current_waypoint)
            self.mission_state = 4
        
        elif self.mission_state == 4:
            """Waypoint 2"""
            self.get_logger().info("Forth waypoint")
            # Imposta secondo waypoint
            self.current_waypoint = np.asfarray([2.5, 2.5, 2], dtype = np.float32)
            self.ENU2NED_vector_converter()
            print(self.current_waypoint)
            self.mission_state = 5

        elif self.mission_state == 5:
            """Landing"""
            self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_NAV_LAND)
            self.mission_state = 6
        
        elif self.mission_state == 6:
            self.timer_offboard.cancel()
            self.get_logger().info("Mission finished")
            self.timer_mission.cancel()
            exit()        
    
    def timer_offboard_cb(self):
        # Funzione richiamata ogni 20ms e invia i seguenti messaggi
        self.publish_offboard_control_mode()
        self.publish_trajectory_setpoint()

    # Arm the vehicle
    def arm(self):
        self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_COMPONENT_ARM_DISARM, 1.0)
        self.get_logger().info("Arm command send")

    # Disarm the vehicle
    # def disarm(self):
    #         self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_COMPONENT_ARM_DISARM, 0.0)
    #         self.get_logger().info("Disarm command send")


    ''' Publish the offboard control mode.
    For this example, only position and altitude controls are active.
    '''

    def publish_offboard_control_mode(self):
        msg = OffboardControlMode()
        msg.position = True
        msg.velocity = False
        msg.acceleration = False
        msg.attitude = False
        msg.body_rate = False
        msg.timestamp = int(Clock().now().nanoseconds / 1000) # time in microseconds
        self.offboard_control_mode_publisher_.publish(msg)

    '''
    Publish a trajectory setpoint
    For this example, it sends a trajectory setpoint to make the
    vehicle hover at 5 meters with a yaw angle of 180 degrees.
    '''

    def publish_trajectory_setpoint(self):
        msg = TrajectorySetpoint()
        msg.position = self.current_waypoint

        #The drone starts with a yaw of pi/2
        msg.yaw = np.pi/2  # [-PI:PI]
        msg.timestamp = int(Clock().now().nanoseconds / 1000) # time in microseconds
            
        self.trajectory_setpoint_publisher_.publish(msg)

    '''
    Publish vehicle commands
        command   Command code (matches VehicleCommand and MAVLink MAV_CMD codes)
        param1    Command parameter 1 as defined by MAVLink uint16 VEHICLE_CMD enum
        param2    Command parameter 2 as defined by MAVLink uint16 VEHICLE_CMD enum
        '''
    def publish_vehicle_command(self, command, param1=0.0, param2=0.0):
        msg = VehicleCommand()
        msg.param1 = param1
        msg.param2 = param2
        msg.command = command  # command ID
        msg.target_system = 1  # system which should execute the command
        msg.target_component = 1  # component which should execute the command, 0 for all components
        msg.source_system = 1  # system sending the command
        msg.source_component = 1  # component sending the command
        msg.from_external = True
        msg.timestamp = int(Clock().now().nanoseconds / 1000) # time in microseconds
        self.vehicle_command_publisher_.publish(msg)

    def ENU2NED_vector_converter(self):

        # Rotation around Z-axis of 90° Degrees
        rotZ = np.asfarray(np.array([[np.cos(np.pi/2), -np.sin(np.pi/2), 0],
                                    [np.sin(np.pi/2), np.cos(np.pi/2), 0], 
                                    [0, 0, 1]]), dtype = np.float32)
        self.current_waypoint = np.dot(rotZ,self.current_waypoint)

        # Rotation around X-axis of 180° Degrees
        rotX = np.asfarray(np.array([[1, 0, 0],
                                    [0, np.cos(np.pi), -np.sin(np.pi)], 
                                    [0, np.sin(np.pi), np.cos(np.pi)]]), dtype = np.float32)
        self.current_waypoint = np.dot(rotX,self.current_waypoint)
    
class GridMap:

    def __init__(self, data, width, height, resolution,
                 center_x, center_y, origin):
        """__init__
        :data: value of each cell either 0 (free) or 100 (occupied)
        :param width: number of grid for width
        :param height: number of grid for height
        :param resolution: grid resolution [m]
        :param center_x: center x position  [m]
        :param center_y: center y position [m]
        :origin: position of starting point
        """

        self.data = data
        self.width = width
        self.height = height
        self.resolution = resolution
        self.center_x = center_x
        self.center_y = center_y
        self.origin = origin

        self.n_data = self.width * self.height

    """ Defines the grid map area to cover"""
    def area_definition(x, y):
        resolution = 1.0 #meters

        # calculate dimensions
        x_max = int((max(x) - min(x))/resolution)
        y_max = int((max(y) - min(y))/resolution)
        
        data = []
        cell_free = 0
        for i in range(y_max):
            for j in range(x_max):
                data.append(cell_free)
        
        width = x_max
        height = y_max
        center_x = 0
        center_y = 0
        origin = [min(x),min(y)]
        gridmap = GridMap(data, width, height, resolution, center_x, center_y, origin)

        return gridmap

    """ Divide the grid map area to cover with multiple CPS"""
    def area_division(grid_map, uav_pose):
    
        # convert swarm pose to grid
        # (i,j) are the coordinates of the position of the drone inside the grid
        print(grid_map.width, grid_map.height, grid_map.origin[:])
        i = (grid_map.width-1) - int((uav_pose[0] - grid_map.origin[0])/grid_map.resolution)
        j = (grid_map.height-1) - int((uav_pose[1] - grid_map.origin[1])/grid_map.resolution)

        return i, j

def main(args=None):
    rclpy.init(args=args)
    print("Starting Offboard Mission...\n")
    offboard_control = OffboardControl()
    rclpy.spin(offboard_control)

    # Edges of the squared surface that needs to be covered
    # Defined in coordinates w.r.t gazebo center
    x = [3, 3, -3, -3]
    y = [3, -3, 3, -3]

    offboard_control.destroy_node()
    rclpy.shutdown()
    print("Clean exit")
    exit(0)

if __name__ == '__main__':
    main()
    
    

File: vicon/src/cpp_swarm/cpp_uav.cpp
#include <px4_msgs/msg/offboard_control_mode.hpp>
#include <px4_msgs/msg/trajectory_setpoint.hpp>
#include <px4_msgs/msg/vehicle_command.hpp>
#include <px4_msgs/msg/vehicle_control_mode.hpp>
#include <rclcpp/rclcpp.hpp>
#include <stdint.h>

#include <chrono>
#include <iostream>

using namespace std::chrono;
using namespace std::chrono_literals;
using namespace px4_msgs::msg;

class OffboardControl : public rclcpp::Node
{
public:

	OffboardControl() : Node("offboard_control")
	{
		rclcpp::Publisher<OffboardControlMode>::SharedPtr offboard_control_mode_publisher_ = this->create_publisher<OffboardControlMode>("/fmu/in/offboard_control_mode", 10);
		rclcpp::Publisher<TrajectorySetpoint>::SharedPtr trajectory_setpoint_publisher_ = this->create_publisher<TrajectorySetpoint>("/fmu/in/trajectory_setpoint", 10);
		rclcpp::Publisher<VehicleCommand>::SharedPtr vehicle_command_publisher_ = this->create_publisher<VehicleCommand>("/fmu/in/vehicle_command", 10);

		int mission_state = 0;
        float current_waypoint[3] = {0.0, 0.0, 0.0};

        void timer_offboard_cb();
        void mission();
        void arm();
        void disarm();
        void publish_trajectory_setpoint();
        void publish_offboard_control_mode();
        void publish_vehicle_command();

        rclcpp::TimerBase::SharedPtr timer_offboard = this->create_wall_timer(0.1s, timer_offboard_cb);
        rclcpp::TimerBase::SharedPtr timer_mission = this->create_wall_timer(10s, mission);

        std::atomic<uint64_t> timestamp_;   //!< common synced timestamped
    }

    void OffboardControl::timer_offboard_cb()
    {
        this->publish_trajectory_setpoint();
        this->publish_offboard_control_mode();
    }
		
    void OffboardControl::mission(){
        if (mission_state == 0){
            // Arming and takeoff to 1st waypoint
            /**
            Python implementation of Offboard Control
            param1 is set to 1 to enable the custom mode.
            param2 is set to 6 to indicate the offboard mode.
            Other options for param2 include:
            param2 = 1 (MANUAL)
            param2 = 2 (ALTCTL)
            param2 = 3 (POSCTL)
            param2 = 4 (AUTO)
            param2 = 5 (ACRO)
            param2 = 6 (OFFBOARD)
            param2 = 7 (STABILIZED)
            param2 = 8 (RATTITUDE)
            **/
            RCLCPP_INFO(this->get_logger(), "Mission Started");
            this->publish_vehicle_command(VehicleCommand::VEHICLE_CMD_DO_SET_MODE, 1, 6);
            mission_state = 1;

        }else if (mission_state == 1)
        {
            this->arm();
            RCLCPP_INFO(this->get_logger(), "Vehicle Armed");
            RCLCPP_INFO(this->get_logger(), "First Waypoint");
            current_waypoint = {-2.5, -2.5, 2}
            current_waypoint = ENU2NED_vector_converter(current_waypoint)
            mission_state = 2

        }else if (mission_state == 2)
        {
            RCLCPP_INFO(this->get_logger(), "Second Waypoint");
            current_waypoint = {-2.5, 2.5, 2}
            current_waypoint = ENU2NED_vector_converter(current_waypoint)
            mission_state = 3
        }else if (mission_state == 3)
        {
            RCLCPP_INFO(this->get_logger(), "Vehicle Landing");
            current_waypoint = {-2.5, 2.5, 2}
            this->publish_vehicle_command(VehicleCommand::VEHICLE_CMD_NAV_LAND);
            mission_state = 4
        }else if (mission_state == 4)
        {
            rclcpp::TimerBase::cancel(this->timer_offboard);
            rclcpp::TimerBase::cancel(this_>timer_mission);
            RCLCPP_INFO(this->get_logger(), "Mission Finished");
        }
            
        }

        /**
         * @brief Send a command to Arm the vehicle
         */
        void arm()
        {
        	publish_vehicle_command(VehicleCommand::VEHICLE_CMD_COMPONENT_ARM_DISARM, 1.0);
        }

        /**
         * @brief Send a command to Disarm the vehicle
         */
        void disarm()
        {
        	publish_vehicle_command(VehicleCommand::VEHICLE_CMD_COMPONENT_ARM_DISARM, 0.0);
        
        	RCLCPP_INFO(this->get_logger(), "Disarm command send");
        }

        /**
         * @brief Publish the offboard control mode.
         *        For this example, only position and altitude controls are active.
         */
        void publish_offboard_control_mode()
        {
        	OffboardControlMode msg{};
        	msg.position = true;
        	msg.velocity = false;
        	msg.acceleration = false;
        	msg.attitude = false;
        	msg.body_rate = false;
        	msg.timestamp = this->get_clock()->now().nanoseconds() / 1000;
        	rclcpp::Publisher<OffboardControlMode>::SharedPtr offboard_control_mode_publisher_->publish(msg);
        }

        /**
         * @brief Publish a trajectory setpoint
         *        For this example, it sends a trajectory setpoint to make the
         *        vehicle hover at 5 meters with a yaw angle of 180 degrees.
         */
        void publish_trajectory_setpoint()
        {
        	TrajectorySetpoint msg{};
        	msg.position = {0.0, 0.0, -5.0};
        	msg.yaw = -3.14; // [-PI:PI]
        	msg.timestamp = this->get_clock()->now().nanoseconds() / 1000;
        	rclcpp::Publisher<OffboardControlMode>::SharedPtr rajectory_setpoint_publisher_->publish(msg);
        }

        /**
         * @brief Publish vehicle commands
         * @param command   Command code (matches VehicleCommand and MAVLink MAV_CMD codes)
         * @param param1    Command parameter 1
         * @param param2    Command parameter 2
         */
        void publish_vehicle_command(uint16_t command, float param1, float param2)
        {
        	VehicleCommand msg{};
        	msg.param1 = param1;
        	msg.param2 = param2;
        	msg.command = command;
        	msg.target_system = 1;
        	msg.target_component = 1;
        	msg.source_system = 1;
        	msg.source_component = 1;
        	msg.from_external = true;
        	msg.timestamp = this->get_clock()->now().nanoseconds() / 1000;
        	rclcpp::Publisher<OffboardControlMode>::SharedPtr vehicle_command_publisher_->publish(msg);
        }
    }
}

int main(int argc, char *argv[])
{
	std::cout << "Starting offboard control node..." << std::endl;
	setvbuf(stdout, NULL, _IONBF, BUFSIZ);
	rclcpp::init(argc, argv);
	rclcpp::spin(std::make_shared<OffboardControl>());

	rclcpp::shutdown();
	return 0;
}